<center>
<img src="https://media.licdn.com/dms/image/C4D22AQEQRRUhbmt-rQ/feedshare-shrink_2048_1536/0/1663956550092?e=1685577600&v=beta&t=m6q43TuX1mIgl0vsRhzpyU4qkn572IlS3T5t0MXR9LM", alt="drawing" width="300">
</center>

# Предобработка данных

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv('cars_df.csv')
print(df.shape)
df.head()

(2982, 16)


Model  \
0        2022 Hyundai Santa Fe Calligraphy   
1                   2016 Lexus IS 300 Base   
2              2017 Chevrolet Volt Premier   
3                      2010 Acura RDX Base   
4  2022 Mercedes-Benz S-Class S 500 4MATIC   

                                                href           Exterior color  \
0  /vehicledetail/41ecf9e2-84c9-48c6-aa5f-36dadd2...              Calypso Red   
1  /vehicledetail/7792ae2c-d42d-439f-8bbd-6d3265e...           Obsidian Black   
2  /vehicledetail/01fe00b8-0863-4f59-a8a8-4dd95da...                    White   
3  /vehicledetail/8cfcbe56-03d7-4c99-bfba-e4b0b3f...  Polished Metal Metallic   
4  /vehicledetail/f5df2c67-2037-4200-8ddf-85d9355...                    White   

  Interior color         Drivetrain  \
0          Black  Front-wheel Drive   
1          Black    All-wheel Drive   
2      Jet Black  Front-wheel Drive   
3          Ebony    All-wheel Drive   
4              –    All-wheel Drive   

                                                 MPG Fuel type  \
0  22–28\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
1  19–26\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
2                                                  –    Hybrid   
3  17–22\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
4                                                  –  Gasoline   

                        Transmission  \
0  8-Speed Automatic with Auto-Shift   
1                  6-Speed Automatic   
2                  1-Speed Automatic   
3                          Automatic   
4                          Automatic   

                                              Engine                VIN  \
0                         2.5L I4 16V PDI DOHC Turbo  5NMS54AL1NH472287   
1                               3.5L V6 24V GDI DOHC  JTHCM1D20G5012454   
2                        1.5L I4 16V GDI DOHC Hybrid  1G1RD6S50HU192847   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  5J8TB1H58AA003867   
4                         3.0L I6 24V GDI DOHC Turbo  W1K6G6DB2NA089310   

     Stock #      Mileage                     Accidents or damage  \
0  NH472287C   13,256 mi.  At least 1 accident or damage reported   
1     RC7373  100,067 mi.  At least 1 accident or damage reported   
2    H17085A   39,032 mi.  At least 1 accident or damage reported   
3      Y3306  123,350 mi.                           None reported   
4   S230029A   10,901 mi.                           None reported   

  1-owner vehicle Personal use only    Price  
0             Yes               Yes  $38,988  
1              No               Yes  $19,747  
2             Yes                No  $25,999  
3              No               Yes  $11,000  
4             Yes                No  $96,750

Подготовленный нами датасет представляет из себя 2982 объявления, опубликованных на американском сайте cars.com, специализирующемся на вторичном рынке автомобилей. Каждое объявление содержит информацию о продаваемой подержанной машине и ее конечную или же желаемую продавцом стоимость.

### Анализ признаков 

Итак, предоставленный для анализа датасет имеет 15 признаков/колонок и 1 целевую переменную, которую мы в финальной части нашего проекта будем предсказывать. Это переменная 'Price' - то есть цена машины. Теперь давайте пройдемся по каждому столбцу нашей таблицы и посмотрим, что в них содержится:
   * **Model** - название модели машины, включающее в себя так же год выпуска и марку автомобиля 
   * **href** - атрибут адреса документа, в нашем случае объявления
   * **Exterior color** - цвет машины 
   * **Interior color** - цвет салона автомобиля 
   * **Drivetrain** - трансмиссия автомобиля, т.е. информация о его ведущие колеса 
   * **MPG** - количество миль, которое автомобиль сможет проехать на одном галлоне топлива
   * **Fuel type** - тип топлива 
   * **Transmission** - информация о коробке передач
   * **Engine** - информация о двигатели автомобиля, содержащая 
      * объем двигателя 
      * конфигурацию 
      * количество клапанов
      * дополнительные характеристки двигателя
   * **VIN** - уникальный идентификационный номер транспортного средства 
   * **Stock #** - складской номер автомобиля
   * **Mileage** - пробег 
   * **Accidents or damage** - наличие аварий у машины 
   * **1-owner vehicle** - информация о количестве владельцев автомобиля, один ли владелец у машины 
   * **Personal use only** - предназначенность машины исключительно для личных целей
   
   
Посмотрев на содержание нашей исходной, еще не обработанной таблицы, мы можем сразу отметить, что такие признаки, как href, VIN и Stock # имеют уникальную для каждой машины информацию и поэтому при дальнейшем построении модели и проверке гипотез они окажутся слишком сложны в анализе и не принесут никакой ценности, поэтому от них стоит сразу избавиться. 

Отметим еще такой признак, как  Transmission, характеризующий коробку передач автомобиля. Запись этого признака каждым из владельцев оказывается индивидуальной, что делает сложным дальнейшую обработку этого столбца и его кодирование в целом.Более того, обльшинство машин на американском рынке имеют автоматическую коробку передач, поэтому исключение из расмотрения этой категории приведет к незначительным изменениям в построении модели и проверке гипотез. Таким образом, как и опысанные выше признаки, мы его удалим.

Так же нетрудно заметить, что некоторые столбцы представляют из себя набор сразу из нескольких признаков, например, Model или Engine. Такие колонки требуют дальнейшей обработки, ведь их компоненты оказываются достаточно значимимы при определении цены машины на вторичном рынке. Поэтому давайте перейдем к их обработке.


In [3]:
# избавимся от слишком уникальных признаков
df.drop(columns=['VIN', 'Stock #', 'href', 'Transmission'], inplace=True)
print(df.shape)
df.head()

(2982, 12)


Model           Exterior color  \
0        2022 Hyundai Santa Fe Calligraphy              Calypso Red   
1                   2016 Lexus IS 300 Base           Obsidian Black   
2              2017 Chevrolet Volt Premier                    White   
3                      2010 Acura RDX Base  Polished Metal Metallic   
4  2022 Mercedes-Benz S-Class S 500 4MATIC                    White   

  Interior color         Drivetrain  \
0          Black  Front-wheel Drive   
1          Black    All-wheel Drive   
2      Jet Black  Front-wheel Drive   
3          Ebony    All-wheel Drive   
4              –    All-wheel Drive   

                                                 MPG Fuel type  \
0  22–28\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
1  19–26\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
2                                                  –    Hybrid   
3  17–22\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
4                                                  –  Gasoline   

                                              Engine      Mileage  \
0                         2.5L I4 16V PDI DOHC Turbo   13,256 mi.   
1                               3.5L V6 24V GDI DOHC  100,067 mi.   
2                        1.5L I4 16V GDI DOHC Hybrid   39,032 mi.   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  123,350 mi.   
4                         3.0L I6 24V GDI DOHC Turbo   10,901 mi.   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
3                           None reported              No               Yes   
4                           None reported             Yes                No   

     Price  
0  $38,988  
1  $19,747  
2  $25,999  
3  $11,000  
4  $96,750

Порой в нашей таблице встречаются пропуски, предтсавляющие из себя не заполненную продавцами информацию о той или иной характеристке машины. Проанализируем их количество в зависимости от признака:

In [4]:
df.isna().sum() #пропуски в данных отсутствуют

Model                  0
Exterior color         0
Interior color         0
Drivetrain             0
MPG                    0
Fuel type              0
Engine                 0
Mileage                0
Accidents or damage    0
1-owner vehicle        0
Personal use only      0
Price                  0
dtype: int64

Однако в наших данных есть иная проблема. Многие продавцы решают по определенным причинам ставить вместо информации значок пропуска, сиволизирующий либо о нежелании распространять информацию, либо же ее отсутствие, а может быть и вовсе пропуск из-за невнимательности. В таких категориальных признаках признаках, как тип топлива и транссмисия, их следует убрать, ведь заменить их нельзя будет, так как мы не знаем какие столь важные технические характериски подойдут именно этой машине.

In [5]:
df = df.drop(df[df['Fuel type'] == '–'].index)
df = df.drop(df[df['Drivetrain'] == '–'].index)

### Обработка колонок

#### Обработка будущих вещественных признаков и целевой переменной

Начнем предобработку данных с преобразования будущих вещественных признаком и самой целевой переменной, а именно приведе их к числовому виду. Такими колонками являются Mileage, Price, MPG.


In [6]:
# определим функцию для цены
def new_price(price):
    return float(price.replace('$', '').replace(',', ''))

df['Price'] = df['Price'].apply(new_price)

In [7]:
df = df.drop(df[df['Mileage'] == 'None reported'].index) # избавимся от аномальных значений, содержащихся в колонке
df = df.drop(df[df['Mileage'] == 'At least 1 accident or damage reported'].index)

In [8]:
# определим функцию для пробега
def new_mileage(mileage):
    return float(mileage.replace('mi.', '').replace(',', ''))

df['Mileage'] = df['Mileage'].apply(new_mileage)

После преобразования цены автомобиля и ее пробега, давайте посмотрим на колонку MPG. Как было описано ранее, данный признак представляет из себя удельный израсход топлива при минимальном и максимальном использовании мощностей автомобиля (нижний и верхний орог интервала соотвественно). Из этого признака мы сможем получить сразу три: MPG_min - минимальный израсход топлива,  MPG_max - максимальный и  MPG_mean - средний израсход топлива автомобилем.

In [9]:
hlp = df['MPG'].str.split('\n\n\n\n', expand=True)  # выделим интервал израсхода топлива от дополнительной информации 
hlp.rename(columns={0: 'MPG_par', 1: 'Data'}, inplace=True)
data = pd.concat([df, hlp], axis=1)
data.drop('MPG', axis=1, inplace=True)

In [10]:
data.head()  # записали интервал в промежуточную колонку MPG_par, а дополнительную информацию в Data

Model           Exterior color  \
0        2022 Hyundai Santa Fe Calligraphy              Calypso Red   
1                   2016 Lexus IS 300 Base           Obsidian Black   
2              2017 Chevrolet Volt Premier                    White   
3                      2010 Acura RDX Base  Polished Metal Metallic   
4  2022 Mercedes-Benz S-Class S 500 4MATIC                    White   

  Interior color         Drivetrain Fuel type  \
0          Black  Front-wheel Drive  Gasoline   
1          Black    All-wheel Drive  Gasoline   
2      Jet Black  Front-wheel Drive    Hybrid   
3          Ebony    All-wheel Drive  Gasoline   
4              –    All-wheel Drive  Gasoline   

                                              Engine   Mileage  \
0                         2.5L I4 16V PDI DOHC Turbo   13256.0   
1                               3.5L V6 24V GDI DOHC  100067.0   
2                        1.5L I4 16V GDI DOHC Hybrid   39032.0   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  123350.0   
4                         3.0L I6 24V GDI DOHC Turbo   10901.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
3                           None reported              No               Yes   
4                           None reported             Yes                No   

     Price MPG_par                                               Data  
0  38988.0   22–28  Based on EPA mileage ratings. Use for comparis...  
1  19747.0   19–26  Based on EPA mileage ratings. Use for comparis...  
2  25999.0       –                                               None  
3  11000.0   17–22  Based on EPA mileage ratings. Use for comparis...  
4  96750.0       –                                               None

In [11]:
k = data['MPG_par'].str.split('–', expand=True)  # отделим теперь границы интервала израсхода топлива
k.rename(columns={0: 'MPG_min', 1: 'MPG_max'}, inplace=True)
data = pd.concat([data,k], axis=1)
data.drop(columns=['MPG_par', 'Data'], inplace=True)

In [12]:
data.head()

Model           Exterior color  \
0        2022 Hyundai Santa Fe Calligraphy              Calypso Red   
1                   2016 Lexus IS 300 Base           Obsidian Black   
2              2017 Chevrolet Volt Premier                    White   
3                      2010 Acura RDX Base  Polished Metal Metallic   
4  2022 Mercedes-Benz S-Class S 500 4MATIC                    White   

  Interior color         Drivetrain Fuel type  \
0          Black  Front-wheel Drive  Gasoline   
1          Black    All-wheel Drive  Gasoline   
2      Jet Black  Front-wheel Drive    Hybrid   
3          Ebony    All-wheel Drive  Gasoline   
4              –    All-wheel Drive  Gasoline   

                                              Engine   Mileage  \
0                         2.5L I4 16V PDI DOHC Turbo   13256.0   
1                               3.5L V6 24V GDI DOHC  100067.0   
2                        1.5L I4 16V GDI DOHC Hybrid   39032.0   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  123350.0   
4                         3.0L I6 24V GDI DOHC Turbo   10901.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
3                           None reported              No               Yes   
4                           None reported             Yes                No   

     Price MPG_min MPG_max  
0  38988.0      22      28  
1  19747.0      19      26  
2  25999.0                  
3  11000.0      17      22  
4  96750.0

Нетрудно заметить, что у нас появились попуски в данных, которые можно заполнить средними показателями по каждому из столбцов:

In [13]:
data.fillna('', inplace=True)  # в ходе преобразований получились наны, уберем их

In [14]:
usl_empty = data['MPG_min'] == ''  # определим условия 
usl_empty_1 = data['MPG_max'] == ''

In [15]:
data['MPG_min'][usl_empty] = 0  # вставим в пустые строки нули, чтобы преобразовать в дальнейшем столбцы в вещественные
data['MPG_max'][usl_empty_1] = 0

In [16]:
data['MPG_min'] = data['MPG_min'].astype(float)
data['MPG_max'] = data['MPG_max'].astype(float)

In [17]:
usl_0 = data['MPG_min'] == 0  # заполним средним значением нулевое
a = data['MPG_min'][~usl_0].mean()
data['MPG_min'][usl_0] = a

In [18]:
usl_0 = data['MPG_max'] == 0  # заполним средним значением нулевое
a = data['MPG_max'][~usl_0].mean()
data['MPG_max'][usl_0] = a

In [19]:
data['MPG_mean'] = (data['MPG_min'] + data['MPG_max']) / 2

In [20]:
data.head()  # посмотрим, что у нас получилось

Model           Exterior color  \
0        2022 Hyundai Santa Fe Calligraphy              Calypso Red   
1                   2016 Lexus IS 300 Base           Obsidian Black   
2              2017 Chevrolet Volt Premier                    White   
3                      2010 Acura RDX Base  Polished Metal Metallic   
4  2022 Mercedes-Benz S-Class S 500 4MATIC                    White   

  Interior color         Drivetrain Fuel type  \
0          Black  Front-wheel Drive  Gasoline   
1          Black    All-wheel Drive  Gasoline   
2      Jet Black  Front-wheel Drive    Hybrid   
3          Ebony    All-wheel Drive  Gasoline   
4              –    All-wheel Drive  Gasoline   

                                              Engine   Mileage  \
0                         2.5L I4 16V PDI DOHC Turbo   13256.0   
1                               3.5L V6 24V GDI DOHC  100067.0   
2                        1.5L I4 16V GDI DOHC Hybrid   39032.0   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  123350.0   
4                         3.0L I6 24V GDI DOHC Turbo   10901.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
3                           None reported              No               Yes   
4                           None reported             Yes                No   

     Price    MPG_min    MPG_max   MPG_mean  
0  38988.0  22.000000  28.000000  25.000000  
1  19747.0  19.000000  26.000000  22.500000  
2  25999.0  20.592777  27.239548  23.916163  
3  11000.0  17.000000  22.000000  19.500000  
4  96750.0  20.592777  27.239548  23.916163

### Обработка будущих категориальных признаков

Посмотрев на такие переменные, как Drivetrain, Exterior color и Interior color, мы можем заметить, что одна и та же информация может быть записана по-разному. В случае трансмиссии это полные и короткие названия типов привода. А вот с цветами солона и машины дело обстоит интереснее. Оба этих признака принимают достаточно большое количество значений. Однако исключение подобных категориальных признаков будет не совсем уместным,ведь цвет играет порой определяющую роль в стоимости машины, поэтому предлагается выделить основные/базовые цвета автомобилей и относить к ним их различные оттенки, с целью упрощения дальнейшего анализа. Если оттенок определить сложно или же будет возникать пропущенное значение в данных, то будет относить такие объекты к одельному типу цвета - Other, то есть другое.

In [21]:
data['Drivetrain'].unique()  # посмотрим, какие значения принимает признак

array(['Front-wheel Drive', 'All-wheel Drive', 'Four-wheel Drive',
       'Rear-wheel Drive', 'FWD', '4WD', 'AWD', 'RWD'], dtype=object)

In [22]:
usl_f = data['Drivetrain'] == 'FWD'  # приведем все значение к полному виду
data.loc[usl_f,'Drivetrain' ] = 'Front-wheel Drive'

usl_a = data['Drivetrain'] == 'AWD'
data.loc[usl_a,'Drivetrain' ] = 'All-wheel Drive'

usl_4 = data['Drivetrain'] == '4WD'
data.loc[usl_4,'Drivetrain' ] = 'Four-wheel Drive'

usl_r = data['Drivetrain'] == 'RWD'
data.loc[usl_r,'Drivetrain' ] = 'Rear-wheel Drive'

In [23]:
# Определим необходимую функцию для работы с цветами
def color_detecktor(color):
    if "Black" in color or "Ebony" in color:
        return "Black"
    elif "White" in color:
        return "White"
    elif "Silver" in color or "Gray" in color:
        return "Silver/Gray"
    elif "Red" in color:
        return "Red"
    elif "Blue" in color:
        return "Blue"
    elif "Green" in color:
        return "Green"
    elif "Yellow" in color:
        return "Yellow"
    elif "Orange" in color:
        return "Orange"
    elif "Brown" in color or "Carmelo" in color or "Tan" in color:
        return "Brown"
    elif "Gold" in color:
        return "Gold"
    elif "Metallic" in color:
        return "Metallic"
    return "Other"

In [24]:
# Обработаем цвета
data['Exterior color'] = data['Exterior color'].apply(color_detecktor)
data['Interior color'] = data['Interior color'].apply(color_detecktor)

### Обработка составных признаков

Теперь займемся разбиение переменных на новые составляющие:

In [25]:
model = df['Model'].str.split(' ', expand=True)  # достанем из столбца модель год выпуска и марку машины
model.rename(columns={0: 'Year', 1: 'Brand'}, inplace=True) 
data = pd.concat([data,model.iloc[:,:2]], axis=1)
data.drop('Model', axis=1, inplace=True)

In [26]:
data['Brand'].unique()  # посмотрим, какие фирмы у нас присутствуют 

array(['Hyundai', 'Lexus', 'Chevrolet', 'Acura', 'Mercedes-Benz', 'Land',
       'INFINITI', 'Chrysler', 'Mazda', 'BMW', 'GMC', 'Toyota',
       'Cadillac', 'Genesis', 'Ford', 'Buick', 'Jeep', 'Audi', 'Nissan',
       'Subaru', 'Lincoln', 'Porsche', 'RAM', 'Kia', 'Lamborghini',
       'McLaren', 'Honda', 'Volkswagen', 'Pontiac', 'Dodge', 'Mitsubishi',
       'Volvo', 'FIAT', 'MINI', 'Bentley', 'Jaguar', 'Aston', 'Maserati',
       'Hummer', 'Alfa', 'Plymouth', 'Scion', 'Rolls-Royce', 'Ferrari',
       'Lotus', 'Polestar', 'Suzuki', 'Saab'], dtype=object)

In [27]:
data.loc[data['Brand'] == 'Land', 'Brand'] = 'Land Rover' # запишем полные названия некоторых марок
data.loc[data['Brand'] == 'Aston', 'Brand'] = 'Aston Martin'
data.loc[data['Brand'] == 'Alfa', 'Brand'] = 'Alfa Romeo'

In [28]:
data['Year'] = data['Year'].astype(int)  # преобразуем переменную year в число

In [29]:
data.head()

Exterior color Interior color         Drivetrain Fuel type  \
0            Red          Black  Front-wheel Drive  Gasoline   
1          Black          Black    All-wheel Drive  Gasoline   
2          White          Black  Front-wheel Drive    Hybrid   
3       Metallic          Black    All-wheel Drive  Gasoline   
4          White          Other    All-wheel Drive  Gasoline   

                                              Engine   Mileage  \
0                         2.5L I4 16V PDI DOHC Turbo   13256.0   
1                               3.5L V6 24V GDI DOHC  100067.0   
2                        1.5L I4 16V GDI DOHC Hybrid   39032.0   
3  2.3L I-4 DOHC, i-VTEC variable valve control, ...  123350.0   
4                         3.0L I6 24V GDI DOHC Turbo   10901.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
3                           None reported              No               Yes   
4                           None reported             Yes                No   

     Price    MPG_min    MPG_max   MPG_mean  Year          Brand  
0  38988.0  22.000000  28.000000  25.000000  2022        Hyundai  
1  19747.0  19.000000  26.000000  22.500000  2016          Lexus  
2  25999.0  20.592777  27.239548  23.916163  2017      Chevrolet  
3  11000.0  17.000000  22.000000  19.500000  2010          Acura  
4  96750.0  20.592777  27.239548  23.916163  2022  Mercedes-Benz

In [30]:
# Вновь определим необходимую функцию для двигателя
def arr_engine_mask(arr):
    if len(arr) == 3:
        if (arr[0][-1] == 'L') & (arr[2][-1] == 'V') & ((arr[1][0] == 'I') | (arr[1][0] == 'V')):
            return arr
    return None

In [31]:
# Обработаем двигателя
engine_str = data['Engine']

engine_arr = engine_str.str.split()
engine_arr = engine_arr.apply(lambda x: x[:3])


engine_df = engine_arr.apply(arr_engine_mask).to_frame()
engine_df = engine_df.rename(columns={'Engine': 'engine_arr'})


cars_df = pd.concat([data, engine_df], join='inner', axis=1)
cars_df.dropna(inplace=True)

In [32]:
# Выделим признаки
capacity = cars_df['engine_arr'].apply(lambda x: x[0])
configuration = cars_df['engine_arr'].apply(lambda x: x[1])
valves = cars_df['engine_arr'].apply(lambda x: x[2])

engine_df = pd.concat(
    [capacity.rename('Capacity'), configuration.rename('Configuration'), valves.rename('Valves')],
    join='inner', axis=1
)

In [33]:
cars_df = pd.concat([cars_df, engine_df], join='inner', axis=1)
cars_df.drop(columns='engine_arr', inplace=True)
cars_df.drop(columns=['Engine'], inplace=True)

In [34]:
# Определим функцию для литров, чтобы преобразовать в число объем двигателя
def new_capacity(capacity):
    return float(capacity.replace('L', ''))

cars_df['Capacity'] = cars_df['Capacity'].apply(new_capacity)

In [35]:
# Определим функцию для количества клапанов
def new_valves(valves):
    return float(valves.replace('V', ''))

cars_df['Valves'] = cars_df['Valves'].apply(new_valves)

In [36]:
cars_df.head()

Exterior color Interior color         Drivetrain Fuel type   Mileage  \
0            Red          Black  Front-wheel Drive  Gasoline   13256.0   
1          Black          Black    All-wheel Drive  Gasoline  100067.0   
2          White          Black  Front-wheel Drive    Hybrid   39032.0   
4          White          Other    All-wheel Drive  Gasoline   10901.0   
5           Blue          Other   Four-wheel Drive  Gasoline   24946.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
4                           None reported             Yes                No   
5                           None reported             Yes               Yes   

     Price    MPG_min    MPG_max   MPG_mean  Year          Brand  Capacity  \
0  38988.0  22.000000  28.000000  25.000000  2022        Hyundai       2.5   
1  19747.0  19.000000  26.000000  22.500000  2016          Lexus       3.5   
2  25999.0  20.592777  27.239548  23.916163  2017      Chevrolet       1.5   
4  96750.0  20.592777  27.239548  23.916163  2022  Mercedes-Benz       3.0   
5  61530.0  17.000000  22.000000  19.500000  2020     Land Rover       3.0   

  Configuration  Valves  
0            I4    16.0  
1            V6    24.0  
2            I4    16.0  
4            I6    24.0  
5            I6    24.0

In [37]:
# Перезапишем столбцы
new_order = ['Brand', 'Year', 'Exterior color', 'Interior color',
             'Drivetrain', 'Fuel type', 'Capacity', 'Configuration', 
             'Valves', 'Accidents or damage', '1-owner vehicle',
            'Personal use only', 'MPG_min', 'MPG_max', 'MPG_mean', 'Mileage', 'Price']

cars_df = cars_df.reindex(columns=new_order)
print(cars_df.shape)
cars_df.head()

(2542, 17)


Brand  Year Exterior color Interior color         Drivetrain  \
0        Hyundai  2022            Red          Black  Front-wheel Drive   
1          Lexus  2016          Black          Black    All-wheel Drive   
2      Chevrolet  2017          White          Black  Front-wheel Drive   
4  Mercedes-Benz  2022          White          Other    All-wheel Drive   
5     Land Rover  2020           Blue          Other   Four-wheel Drive   

  Fuel type  Capacity Configuration  Valves  \
0  Gasoline       2.5            I4    16.0   
1  Gasoline       3.5            V6    24.0   
2    Hybrid       1.5            I4    16.0   
4  Gasoline       3.0            I6    24.0   
5  Gasoline       3.0            I6    24.0   

                      Accidents or damage 1-owner vehicle Personal use only  \
0  At least 1 accident or damage reported             Yes               Yes   
1  At least 1 accident or damage reported              No               Yes   
2  At least 1 accident or damage reported             Yes                No   
4                           None reported             Yes                No   
5                           None reported             Yes               Yes   

     MPG_min    MPG_max   MPG_mean   Mileage    Price  
0  22.000000  28.000000  25.000000   13256.0  38988.0  
1  19.000000  26.000000  22.500000  100067.0  19747.0  
2  20.592777  27.239548  23.916163   39032.0  25999.0  
4  20.592777  27.239548  23.916163   10901.0  96750.0  
5  17.000000  22.000000  19.500000   24946.0  61530.0

Таким образом, у нас получилась итоговая таблица, содержащая следующие признаки:

   * **Brand** - марка автомбиля *(категориальный признак)*
   * **Year** - год выпуска машины *(вещественный признак)*
   * **Exterior color** - цвет машины *(категориальный признак)*
   * **Interior color** - цвет салона *(категориальный признак)*
   * **Drivetrain** - трансмиссия автомобиля, т.е. информация о его ведущие колеса  *(категориальный признак)*
   * **Fuel type** - тип топлива *(категориальный признак)*
   * **Capacity** - объем двигателя, литры *(вещественный признак)*
   * **Configuration** - конфигурация автомобиля *(категориальный признак)*
   * **Valves** - количество клапанов, штуки *(вещественный признак)*
   * **Accidents or damage** - наличие аварий у машины *(бинарный признак)*
   * **1-owner vehicle** - информация о количестве владельцев автомобиля, один ли владелец у машины *(бинарный признак)*
   * **Personal use only** - предназначенность машины исключительно для личных целей *(бинарный признак)*
   * **MPG_min** - количество миль, которое автомобиль сможет проехать на одном галлоне топлива при минимальном использовании *(вещественный признак)*
   * **MPG_max** - количество миль, которое автомобиль сможет проехать на одном галлоне топлива при максимальном использовании *(вещественный признак)*
   * **MPG_mean** - количество миль, которое автомобиль сможет проехать на одном галлоне топлива при среднем использовании *(вещественный признак)*
   * **Mileage** - пробег, мили *(вещественный признак)*
   
И одна целевая переменная **Price** - цена автомобиля, доллары.

In [38]:
# Сохраним итоговую таблицу
cars_df.to_csv('cars_df_interim.csv', index=False)